README.md
# annuaire
Nouvelle version de l'annuaire (SSO) Tela Botanica basée sur Wordpress

## installation
composer install --no-dev

## configuration
```
cd config
cp config.default.json config.json
cp service.default.json service.json
cp clef-auth.defaut.ini clef-auth.ini
```
### service.json
 - **domain_root** : racine du domaine, sans le protocole (ex: "beta.tela-botanica.org")
 - **base_uri** : URI de base du service (ex: "/service:annuaire"), dépend des redirections choisies
 - **first_resource_separator** : permier caractère attendu après base_uri (ex: "/" ou ":")
 - **auth** : voir la documentation dans le fichier

### config.json
 - **adapter** : implémentation du stockage, laisser "AnnuaireWPAPI" (seule implémentation fonctionnelle pour l'instant)
 - **adapters**
  - **AnnuaireWPAPI**
    - **chemin_wp** : chemin de l'installation de Wordpress (ex: "/home/user/www/test")
    - **prefixe_tables** : préfixe des tables Wordpress pour l'installation ciblée (ex: "wp_")
  - **auth**
    - **mdp_magique_hache** : si un mot de passe haché en MD5 est placé ici, ce mot de passe permettra d'acéder à tous les comptes

### clef-auth.ini
Saisir une suite d'au moins 16 caractères dans ce fichier.

Cette clef sert à signer les jetons du service d'authentification.

En cas de problème de sécurité (fuite de jeton admin par ex.), modifier ce fichier invalidera tous les jetons existants.

## gestion des rôles SSO
Les services **utilisateur** et **auth** renvoient la liste des rôles affectés à l'utilisateur.

Pour cumuler plusieurs rôles (nécessaire pour gérer les permissions SSO), nécessite le plugin Wordpress "Multiple Roles" : https://fr.wordpress.org/plugins/multiple-roles/

Charge à l'administrateur WP d'ajouter des rôles.


CHANGELOG
2016-11-17 : v2.0-Bouvier
 * Passage à Git
 * Implémentation de l'annuaire sur Wordpress / BuddyPress
 * Ajout de la synchronisation de comptes partenaires (modification si déjà existant)
 * Abandon de toutes les interfaces graphiques (maintenant géré par Wordpress)
 * Abandon du framework TB au profit des paquets Composer TB
 * Implémentation du service avec un patron de conception Adaptateur pour remplacer
   facilement la couche de stockage au besoin
 * Implémentation du système de rôles centralisé pour le SSO, basé sur les rôles WP
 * Implémentation du système de groupes, basé sur les groupes BP
 * Ajout d'un script pour forger des jetons admin longue durée
 * Ajout de tests unitaires

2016-03-01 : v1.5-Bélier
 * Première version de l'authentification SSO par un partenaire - implémenté basiquement pour Pl@ntNet
 * Division de la Yougoslavie en Serbie et Monténégro (12 ans après...)
 * Caractérisation (mal faite !) de Chypre, qui était incluse dans la Grèce (0 pointé en géographie)
 * Service Auth : acceptation du nom d'utilisateur en plus du courriel, pour Pl@ntNet
 * Bricolage cracra d'AnnuaireModele pour faciliter les inscriptions lors d'une connexion à un partenaire (@TODO faire mieux un jour)
 * La casse des adresses courriel est maintenant ignorée
 * Auth => redirection  : ajout du parametre Authorization même si le jeton est null; suppression du header qui ne sert à rien (reçu par le navigateur mais pas renvoyé à l'URL de redirection !)
 * Implémentation d'un mot de passe universel pour le SSO
 * Dépatchage du patch passoire : le patch CGI permettait à tout le monde de se logger avec n'importe quel mdp :)
 * Amélioration gestion du nom wiki
 * Patch : si PHP fonctionne en CGI, pas moyen (en tout cas pas facile - testé une combine mais elle marche pas) d'utiliser HTTP Basic Auth (la merde !); on détecte donc si on est en CGI et dans ce cas on désactive l'authentification - open bar !
 * Mode synchrone avec redirection si une url de redirection est précisée pour le sso
 * Utilisateur : suppression méthode temporaire getIdentiteConnectee()
 * Auth : connexion() retourne maintenant le nom Wiki
 * Utilisateur : formaterNomWiki() tient compte du pseudo
 * Auth : verifierjeton prend maintenant en compte le header Authorization + correction doc
 * Auth : une tentative de connexion infructueuse détruit le cookie, pour plus de cohérence
 

2015-05-26 : v1.4-Baleine - ATTENTION, nécessite PHP >= 5.3
 * Service Auth : première version du SSO à base de jetons JWT
 * Ajout lib firebase/php-jwt
 * Ajout composer.json et autoloader pour firebase/php-jwt
 * Ajout de l'identifiant et de la durée du jeton dans la réponse JSON
 * .htaccess : ajout d'une directive pour utiliser PHP 55 en CGI sur Sequoia (pas très générique !)
 * Auth : ajout de "secure" au cookie
 * Service auth::identité : lecture du jeton dans le header "Authorization", après le cookie et avant le paramètre
 * Protection contres indices de tableau inexistants
 * Amélioration des sorties du programme
 * Affichage du log sur plusieurs lignes pour le retour du script
 * Script de traitement des emails : essai d'élargissement du wait_timeout pour éviter les erreurs "MySQL has gone away"
 * Ajout d'une vérification d'unicité du mail lors de la validation d'inscription
 * Ajout d'un nom d'utilisateur en haut de page dans l'onglet "dons"
 * Changement de logos et favicons : liens vers resources.tela-botanica.org
 * Ajout du service getIdentiteCompleteParCourriel() et de la possibilité de renvoyer du XML, pour rétrocompatibilité avec le service coel:CoelUtilisateur qui utilisait eflore_chatin
 * Ajout d'une méthode d'authentification HTTP Basic
 * Changement signature David vers signature Daniel
 * Modification de getInfosParId(s) pour traiter plusieurs ids à la fois
 * Correction cartes Afrique, Reçu de dons, cartoPoint dans le profil utilisateur
 * Ajout de fonctions de pseudo sso en attendant le vrai

2014-09-29 : v1.3-Balance : utilisation de scripts différés pour gérer les grandes quantités de mail
2014-02-18 : v1.2-Autel : intégration de l'ensemble des modifications depuis le 2010-08-12
2010-08-12 : v1.1-Andromede : corrections 
2010-08-03 : v1.0-Aigle : version initiale 

services/Auth.php
<?php
/**
 * Tentative de service d'authentification / SSO bien organisé
 * SSO - Mettez un tigre dans votre annuaire !
 * @author mathias
 * © Tela Botanica 2015 - 2016
 */
class Auth extends BaseRestServiceTB {

	/** Chemin du fichier de clef */
	const CHEMIN_CLEF_AUTH = "config/clef-auth.ini";

	/** Clef utilisée pour signer les jetons JWT */
	private $clef;

	/** Si true, refusera une connexion non-HTTPS */
	protected $forcerSSL = true;

	/** Durée en secondes du jeton (doit être faible en l'absence de mécanisme d'invalidation) */
	protected $dureeJeton = 900;

	/** Durée en secondes du cookie */
	protected $dureeCookie = 31536000; // 3600 * 24 * 365

	/** Nom du cookie */
	protected $nomCookie = "this_is_not_a_good_cookie_name";

	/** Domaine du cookie - lire la doc de set_cookie() */
	protected $domaineCookie = null;

	/** Si true, passera secure=true à setCookie; le cookie sera reçu en HTTPS seulement */
	protected $cookieSecurise = true;

	/** Bibliothèque de gestion des utilisateurs */
	protected $annuaire;

	public function __construct($config, $annuaire) {
		parent::__construct($config);
		// Auth est un sous-service : suppression de la première ressource,
		// qui a servi au service parent (Annuaire) pour aiguiller ici
		// @TODO faire une classe de sous-service plus astucieuse un jour
		array_shift($this->resources);

		$this->clef = file_get_contents(self::CHEMIN_CLEF_AUTH);
		if (strlen($this->clef) < 16) {
			throw new Exception("Clef trop courte - placez une clef d'au moins 16 caractères dans configurations/clef-auth.ini");
		}

		$this->forcerSSL = ($this->config['auth']['forcer_ssl'] == "1");
		$this->dureeJeton = $this->config['auth']['duree_jeton'];
		$this->dureeCookie = $this->config['auth']['duree_cookie'];
		$this->nomCookie = $this->config['auth']['nom_cookie'];
		$this->cookieSecurise = $this->config['auth']['cookie_securise'];
		if (! empty($this->config['auth']['domaine_cookie'])) {
			$this->domaineCookie = $this->config['auth']['domaine_cookie'];
		}

		// lib annuaire
		$this->annuaire = $annuaire;
	}

	/**
	 * Retourne la bibliothèque annuaire (pour les classes partenaires)
	 */
	public function getAnnuaire() {
		return $this->annuaire;
	}

	/**
	 * Notice d'utilisation succincte
	 * @TODO essayer de choisir entre anglais et français
	 */
	protected function infosService() {
		$protocole = $this->isHTTPS ? 'https://' : 'http://';
		$uri = $protocole . $this->config['domain_root'] . $this->config['base_uri'];
		$uri = $uri . $this->firstResourceSeparator . "auth/";

		$infos = array(
			'service' => 'TelaBotanica/annuaire/auth',
			'methodes' => array(
				'connexion' => array(
					"uri" => $uri . "connexion",
					"parametres" => array(
						"login" => "adresse email (ex: name@domain.com)",
						"password" => "mot de passe",
						"partner" => "nom du partenaire (ex: plantnet)"
					),
					"alias" => $uri . "login",
					"description" => "connexion avec login et mot de passe; renvoie un jeton et un cookie " . $this->nomCookie
				),
				'deconnexion' => array(
					"uri" => $uri . "deconnexion",
					"parametres" => null,
					"alias" => $uri . "logout",
					"description" => "déconnexion; renvoie un jeton null et supprime le cookie " . $this->nomCookie
				),
				'identite' => array(
					"uri" => $uri . "identite",
					"parametres" => array(
						"token" => "jeton JWT (facultatif)",
					),
					"alias" => array(
						$uri . "identity",
						$uri . "rafraichir",
						$uri . "refresh"
					),
					"description" => "confirme l'authentification et la session; rafraîchit le jeton fourni (dans le cookie " . $this->nomCookie . ", le header Authorization ou en paramètre)"
				),
				'verifierjeton' => array(
					"uri" => $uri . "verifierjeton",
					"parametres" => array(
						"token" => "jeton JWT",
					),
					"alias" => $uri . "verifytoken",
					"description" => "retourne true si le jeton fourni en paramètre ou dans le header Authorization est valide, une erreur sinon"
				)
			)
		);
		$this->sendJson($infos);
	}

	public function get() {
		if (count($this->resources) == 0) {
			// sans élément d'URL, on sort du service
			$this->infosService();
		} else {
			// Achtétépéèch portouguech lolch
			$this->verifierSSL();
			// le premier paramètre d'URL définit la méthode (non-magique)
			switch ($this->resources[0]) {
				case 'login':
				case 'connexion':
					$this->connexion();
					break;
				case 'logout':
				case 'deconnexion':
					$this->deconnexion();
					break;
				case 'identity':
				case 'identite':
				case 'rafraichir':
				case 'refresh':
					$this->identite();
					break;
				case 'verifytoken':
				case 'verifierjeton':
					$this->verifierJeton();
					break;
				case 'info':
				default:
					$this->infosService();
			
			}
		}
	}

	/**
	 * Lors d'un POST avec au moins une donnée dans le body (data);
	 * les paramètres GET sont ignorés
	 * @TODO faire un point d'entrée POST qui renvoie vers les méthodes GET
	 * 
	 * @param array $ressources les éléments d'URL
	 * @param array $pairs les paramètres POST
	 */
	public function updateElement($ressources, $pairs) {
		//echo "update element\n";
		$this->nonImplemente();
	}

	/**
	 * Lors d'un PUT (les éléments d'URL sont ignorés) ou d'un POST avec au moins
	 * un élément d'URL; dans tous les cas les paramètres GET sont ignorés
	 * 
	 * @param array $pairs les paramètres POST
	 */
	public function createElement($pairs) {
		//echo "create element\n";
		$this->nonImplemente();
	}

	/**
	 * Lors d'un DELETE avec au moins un élément d'URL
	 * @TODO utiliser pour invalider un jeton (nécessite stockage)
	 * 
	 * @param array $ressources les éléments d'URL
	 */
	public function deleteElement($ressources) {
		//echo "delete element\n";
		$this->nonImplemente();
	}

	/**
	 * Vérifie l'identité d'un utilisateur à partir de son courriel et son
	 * mot de passe ou d'un cookie; lui accorde un jeton et un cookie si
	 * tout va bien, sinon renvoie une erreur et détruit le cookie
	 * @WARNING si vous n'utilisez pas urlencode() pour fournir le mot de passe,
	 * le caractère "&" posera problème en GET
	 */
	protected function connexion() {
		$login = $this->getParam('login');
		$password = $this->getParam('password', null);
		$partenaire = $this->getParam('partner');
		if ($login == '' || $password == '') {
			$this->sendError("parameters <login> and <password> required");
		}
		$acces = false;
		// connexion à un partenaire ?
		$infosPartenaire = array();
		if ($partenaire != '') {
			$classeAuth = "AuthPartner" . ucfirst(strtolower($partenaire));
			try {
				$fichierClasse = getcwd() . "/services/auth/$classeAuth.php"; // @TODO vérifier si getcwd() est fiable dans ce cas
				if (! file_exists($fichierClasse)) {
					$this->sendError("unknown partner '$partenaire'");
				}
				require $fichierClasse;
				$authPartenaire = new $classeAuth($this, $this->config);
				// authentification par le partenaire
				$acces = $authPartenaire->verifierAcces($login, $password);
				if ($acces === true) {
					// copie des infos dans l'annuaire si besoin
					$authPartenaire->synchroniser();
				}
				// détails à ajouter au jeton local
				$infosPartenaire['partenaire'] = $partenaire;
				$infosPartenaire['jetonPartenaire'] = $authPartenaire->getJetonPartenaire();
				// remplacement du login par le courriel (chez certains partenaires,
				// le login peut ne pas être un courriel
				$login = $authPartenaire->getCourriel();
			} catch(Exception $e) {
				$this->sendError($e->getMessage(), 500);
			}
		} else {
			// authentification locale
			$acces = $this->annuaire->verifierAcces($login, $password);
		}
		if ($acces === false) {
			$this->detruireCookie();
			// redirection si demandée - se charge de sortir du script en cas de succès
			$this->rediriger();
			// si la redirection n'a pas eu lieu
			$this->sendError("authentication failed", 401);
		}
		// infos utilisateur
		$infos = $this->annuaire->infosParCourriels($login);
		// getIdentiteParCourriel retourne toujours le courriel comme clef de tableau en lowercase
		$login = strtolower($login);
		if (count($infos) == 0 || empty($infos[$login])) {
			// redirection si demandée - se charge de sortir du script en cas de succès
			$this->rediriger();
			// si la redirection n'a pas eu lieu
			$this->sendError("could not get user info");
		}
		$infos = $infos[$login];
		// date de dernière modification du profil
		$dateDerniereModif = $this->annuaire->getDateDerniereModifProfil($infos['id'], true);
		$infos['dateDerniereModif'] = $dateDerniereModif;
		// infos partenaire
		$infos = array_merge($infos, $infosPartenaire);
		// création du jeton
		$jwt = $this->creerJeton($login, $infos);
		// création du cookie
		$this->creerCookie($jwt);
		// redirection si demandée - se charge de sortir du script en cas de succès
		$this->rediriger($jwt);
		// envoi
		$this->sendJson(array(
			"session" => true,
			"token" => $jwt,
			"duration" => intval($this->dureeJeton),
			"token_id" => $this->nomCookie,
			"last_modif" => $infos['dateDerniereModif']
		));
	}

	/**
	 * Détruit le cookie et renvoie un jeton vide ou NULL - le client
	 * devrait toujours remplacer son jeton par celui renvoyé par les
	 * méthodes de l'annuaire
	 */
	protected function deconnexion() {
		// suppression du cookie
		$this->detruireCookie();
		// envoi d'un jeton null
		$jwt = null;
		// redirection si demandée - se charge de sortir du script en cas de succès
		$this->rediriger();
		// si la redirection n'a pas eu lieu
		$this->sendJson(array(
				"session" => false,
				"token" => $jwt,
				"token_id" => $this->nomCookie
		));
	}

	/**
	 * Renvoie un jeton rafraîchi (durée de validité augmentée de $this->dureeJeton
	 * si l'utilisateur est reconnu comme détenteur d'une session active (cookie valide,
	 * header HTTP "Authorization" ou jeton valide); renvoie une erreur si le cookie
	 * et/ou le jeton sont expirés;
	 * cela permet en théorie de forger des cookies avec des jetons expirés pour se les
	 * faire rafraîchir frauduleusement, mais le canal HTTPS fait qu'un client ne peut
	 * être en possession que de ses propres jetons... au pire on peut se faire prolonger
	 * à l'infini même si on n'est plus inscrit à l'annuaire... @TODO faire mieux un jour
	 * Priorité : cookie > header "Authorization" > paramètre "token" @TODO vérifier cette
	 * stratégie, l'inverse est peut-être plus malin
	 */
	protected function identite() {
		$cookieAvecJetonValide = false;
		$jetonRetour = null;
		$erreur = '';
		// lire cookie
		if (isset($_COOKIE[$this->nomCookie])) {
			$jwt = $_COOKIE[$this->nomCookie];
			try {
				// rafraîchir jeton quelque soit son état - "true" permet
				// d'ignorer les ExpiredException (on rafraîchit le jeton
				// expiré car le cookie est encore valide)
				$jetonRetour = $this->rafraichirJeton($jwt, true);
				// on ne tentera pas de lire un jeton fourni en paramètre
				$cookieAvecJetonValide = true;
			} catch (Exception $e) {
				// si le rafraîchissement a échoué (jeton invalide - hors expiration - ou vide)
				// on ne fait rien et on tente la suite (jeton fourni hors cookie ?)
				$erreur = "invalid token in cookie";
			}
		}
		// si le cookie n'existait pas ou ne contenait pas un jeton
		if (! $cookieAvecJetonValide) {
			// lire jeton depuis header ou paramètre
			$jwt = $this->lireJetonDansHeader();
			if ($jwt == null) {
				// dernière chance
				$jwt = $this->getParam('token');
			}
			// toutes les possibilités ont été essayées
			if ($jwt != null) {
				try {
					// rafraîchir jeton si non expiré
					$jetonRetour = $this->rafraichirJeton($jwt);
				} catch (Exception $e) {
					// si le rafraîchissement a échoué (jeton invalide, expiré ou vide)
					$erreur = "invalid or expired token in Authorization header or parameter <token>";
				}
			} else {
				// pas de jeton valide passé en paramètre
				$erreur = ($erreur == "" ? "no token or cookie" : "invalid token in cookie / invalid or expired token in Authorization header or parameter <token>");
			}
		}
		// redirection si demandée - se charge de sortir du script en cas de succès
		$this->rediriger($jetonRetour);
		// renvoi jeton
		if ($jetonRetour === null) {
			$this->sendError($erreur);
		} else {
			$this->sendJson(array(
					"session" => true,
					"token" => $jetonRetour,
					"duration" => intval($this->dureeJeton),
					"token_id" => $this->nomCookie
			));
		}
	}

	/**
	 * Si $_GET['redirect_url'] est non-vide, redirige vers l'URL qu'il contient et sort du programme;
	 * sinon, ne fait rien et passe la main
	 * 
	 * @param string $jetonRetour jeton JWT à passer à l'URL de destination
	 * en GET; par défaut null
	 */
	protected function rediriger($jetonRetour=null) {
		if (!empty($_GET['redirect_url'])) {
			// dans le cas où une url de redirection est précisée,
			// on précise le jeton dans le get
			$url_redirection = $_GET['redirect_url'];

			// même si le jeton est vide, on ajoute un paramètre GET Authorization
			// pour spécifier à la cible qu'on a bien traité sa requête - permet
			// aussi de gérer les déconnexions en renvoyant un jeton vide
			$separateur = (parse_url($url_redirection, PHP_URL_QUERY) == NULL) ? '?' : '&';
			$url_redirection .= $separateur.'Authorization='.$jetonRetour;

			// retour à l'envoyeur !
			header('Location: '.$url_redirection);
			exit;
		}
	}

	/**
	 * Vérifie si un jeton est valide; retourne true si oui, une erreur avec
	 * des détails si non;
	 * Priorité : header "Authorization" > paramètre "token"
	 */
	protected function verifierJeton() {
		// vérifie que le jeton provient bien d'ici,
		// et qu'il est encore valide (date)
		$jwt = $this->lireJetonDansHeader();
		if ($jwt == null) {
			$jwt = $this->getParam('token');
			if ($jwt == '') {
				$this->sendError("parameter <token> or Authorization header required");
			}
		}
		try {
			$jeton = JWT::decode($jwt, $this->clef, array('HS256'));
			$jeton = (array) $jeton;
		} catch (Exception $e) {
			$this->sendError($e->getMessage());
			exit;
		}
		$this->sendJson(true);
	}

	/**
	 * Reçoit un jeton JWT, et s'il est non-vide ("sub" != null), lui redonne
	 * une période de validité de $this->dureeJeton; si $ignorerExpiration
	 * vaut true, rafraîchira le jeton même s'il a expiré
	 * (attention à ne pas appeler cette méthode n'importe comment !);
	 * jette une exception si le jeton est vide, mal signé ou autre erreur,
	 * ou s'il a expiré et que $ignorerExpiration est différent de true
	 * 
	 * @param string $jwt le jeton JWT
	 * @return string le jeton rafraîchi
	 */
	protected function rafraichirJeton($jwt, $ignorerExpiration=false) /* throws Exception */ {
		$infos = array();
		// vérification avec lib JWT
		try {
			$infos = JWT::decode($jwt, $this->clef, array('HS256'));
			$infos = (array) $infos;
		} catch (ExpiredException $e) {
			if ($ignorerExpiration === true) {
				// on se fiche qu'il soit expiré
				// décodage d'un jeton expiré 
				// @WARNING considère que la lib JWT jette ExpiredException en dernier (vrai 12/05/2015),
				// ce qui signifie que la signature et le domaine sont tout de même valides - à surveiller !
				$infos = $this->decoderJetonManuellement($jwt);
			} else {
				// on renvoie l'exception plus haut
				throw $e;
			}
		}
		// vérification des infos
		if (empty($infos['sub'])) {
			// jeton vide (wtf?)
			throw new Exception("empty token (no <sub>)");
		}
		// rafraîchissement
		$infos['exp'] = time() + $this->dureeJeton;
		$jwtSortie = JWT::encode($infos, $this->clef);

		return $jwtSortie;
	}

	/**
	 * Décode manuellement un jeton JWT, SANS VÉRIFIER SA SIGNATURE OU
	 * SON DOMAINE ! @WARNING ne pas utiliser hors du cas d'un jeton
	 * correct (vérifié avec la lib JWT) mais expiré !
	 * Public car utilisé par les classes AuthPartner (@TODO stratégie à valider)
	 * @param string $jwt un jeton vérifié comme valide, mais expiré
	 */
	public function decoderJetonManuellement($jwt) {
		$parts = explode('.', $jwt);
		$payload = $parts[1];
		$payload = base64_decode($payload);
		$payload = json_decode($payload, true);

		return $payload;
	}

	/**
	 * Crée un jeton JWT signé avec la clef
	 * 
	 * @param mixed $sub subject: l'id utilisateur du détenteur du jeton si authentifié, null sinon
	 * @param array $donnees les données à ajouter au jeton (infos utilisateur)
	 * @param string $exp la date d'expiration du jeton, par défaut la date actuelle plus $this->dureeJeton
	 * 
	 * @return string un jeton JWT signé
	 */
	protected function creerJeton($sub, $donnees=array(), $exp=null) {
		if ($exp === null) {
			$exp = time() + $this->dureeJeton;
		}
		$jeton = array(
			"iss" => "https://www.tela-botanica.org",
			"token_id" => $this->nomCookie,
			//"aud" => "http://example.com",
			"sub" => $sub,
			"iat" => time(),
			"exp" => $exp,
			//"nbf" => time() + 60,
			"scopes" => array("tela-botanica.org")
		);
		if (! empty($donnees)) {
			$jeton = array_merge($jeton, $donnees);
		}
		$jwt = JWT::encode($jeton, $this->clef);

		return $jwt;
	}

	/**
	 * Essaye de trouver un jeton JWT non vide dans l'entête HTTP $nomHeader (par
	 * défaut "Authorization")
	 * 
	 * @param string $nomHeader nom de l'entête dans lequel chercher le jeton
	 * @return String un jeton JWT ou null
	 */
	protected function lireJetonDansHeader($nomHeader="Authorization") {
		$jwt = null;
		$headers = apache_request_headers();
		if (isset($headers[$nomHeader]) && ($headers[$nomHeader] != "")) {
			$jwt = $headers[$nomHeader];
		}
		return $jwt;
	}

	/**
	 * Crée un cookie de durée $this->dureeCookie, nommé $this->nomCookie et
	 * contenant $valeur
	 * 
	 * @param string $valeur le contenu du cookie (de préférence un jeton JWT)
	 */
	protected function creerCookie($valeur) {
		setcookie($this->nomCookie, $valeur, time() + $this->dureeCookie, '/', $this->domaineCookie, $this->cookieSecurise);
	}

	/**
	 * Renvoie le cookie avec une valeur vide et une date d'expiration dans le
	 * passé, afin que le navigateur le détruise au prochain appel
	 * @TODO envisager l'envoi d'un jeton vide plutôt que la suppression du cookie
	 * 
	 * @param string $valeur la valeur du cookie, par défaut ""
	 */
	protected function detruireCookie() {
		setcookie($this->nomCookie, "", -1, '/', $this->domaineCookie, $this->cookieSecurise);
		// mode transition: supprime l'ancien cookie posé sur "www.tela-botanica.org" sans quoi on ne peut plus se déconnecter!
		// @TODO supprimer au bout d'un moment
		setcookie($this->nomCookie, "", -1, '/', null, $this->cookieSecurise);
	}

	/**
	 * Message succinct pour méthodes / actions non implémentées
	 */
	protected function nonImplemente() {
		$this->sendError("not implemented");
	}
	
	/**
	 * Si $this->forcerSSL vaut true, envoie une erreur et termine le programme si SSL n'est pas utilisé
	 */
	protected function verifierSSL() {
		if ($this->forcerSSL === true && ! $this->isHTTPS) {
			$this->sendError("HTTPS required");
			exit;
		}
	}
}

/**
 * Mode moderne pour PHP < 5.4
 * @TODO supprimer lors du démantèlement de Sequoia
 */
if (!function_exists('http_response_code')) {
	function http_response_code($code = NULL) {
		if ($code !== NULL) {
			switch ($code) {
				case 100: $text = 'Continue'; break;
				case 101: $text = 'Switching Protocols'; break;
				case 200: $text = 'OK'; break;
				case 201: $text = 'Created'; break;
				case 202: $text = 'Accepted'; break;
				case 203: $text = 'Non-Authoritative Information'; break;
				case 204: $text = 'No Content'; break;
				case 205: $text = 'Reset Content'; break;
				case 206: $text = 'Partial Content'; break;
				case 300: $text = 'Multiple Choices'; break;
				case 301: $text = 'Moved Permanently'; break;
				case 302: $text = 'Moved Temporarily'; break;
				case 303: $text = 'See Other'; break;
				case 304: $text = 'Not Modified'; break;
				case 305: $text = 'Use Proxy'; break;
				case 400: $text = 'Bad Request'; break;
				case 401: $text = 'Unauthorized'; break;
				case 402: $text = 'Payment Required'; break;
				case 403: $text = 'Forbidden'; break;
				case 404: $text = 'Not Found'; break;
				case 405: $text = 'Method Not Allowed'; break;
				case 406: $text = 'Not Acceptable'; break;
				case 407: $text = 'Proxy Authentication Required'; break;
				case 408: $text = 'Request Time-out'; break;
				case 409: $text = 'Conflict'; break;
				case 410: $text = 'Gone'; break;
				case 411: $text = 'Length Required'; break;
				case 412: $text = 'Precondition Failed'; break;
				case 413: $text = 'Request Entity Too Large'; break;
				case 414: $text = 'Request-URI Too Large'; break;
				case 415: $text = 'Unsupported Media Type'; break;
				case 500: $text = 'Internal Server Error'; break;
				case 501: $text = 'Not Implemented'; break;
				case 502: $text = 'Bad Gateway'; break;
				case 503: $text = 'Service Unavailable'; break;
				case 504: $text = 'Gateway Time-out'; break;
				case 505: $text = 'HTTP Version not supported'; break;
				case 666: $text = 'Couscous overheat'; break;
				default:
					exit('Unknown http status code "' . htmlentities($code) . '"');
					break;
			}

			$protocol = (isset($_SERVER['SERVER_PROTOCOL']) ? $_SERVER['SERVER_PROTOCOL'] : 'HTTP/1.0');
			header($protocol . ' ' . $code . ' ' . $text);
			$GLOBALS['http_response_code'] = $code;
		} else {
			$code = (isset($GLOBALS['http_response_code']) ? $GLOBALS['http_response_code'] : 200);
		}
		return $code;
	}
}
